### PR TITLE
Istanbul: EIP-152 Blake2 F function

### LINKED ISSUE #568: [Istanbul] EIP-2129/152: Add Blake2b F precompile
Part of #501.

Depends on https://github.com/ethereumjs/ethereumjs-util/issues/203.

The hash function implemented above needs to be added here as a new precompile for Istanbul.

@s1na has opted in to do the implementation.

### PR DESCRIPTION
**Pull request by s1na**
Work in progress to fix #568

I took the blake2 logic from [blakejs](https://github.com/dcposch/blakejs). For now I put it right above the precompile to make testing easier. I'm not sure where we want to maintain this code. Ideas @holgerd77?

Also used test cases in geth, one which does 8 million rounds is failing. I was trying to use the cases in https://github.com/ethereum/tests/pull/619 but they all fail. I think it might be due to us not having implemented EIP-2200 yet.

Also the linter insisted on putting every element of `SIGMA8` and `IV` on their own lines (adding 100 extra LoC), which I find ugly :stuck_out_tongue:

### CODE REVIEW

File: lib/evm/precompiles/09-blake2f.ts [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    my[i] = B2B_GET32(m, 4 * i)
+  }
+
+  // twelve rounds of mixing
+  // uncomment the DebugPrint calls to log the computation
+  // and match the RFC sample documentation
+  // util.debugPrint('          m[16]', m, 64)
+  for (i = 0; i < rounds; i++) {
+    // util.debugPrint('   (i=' + (i < 10 ? ' ' : '') + i + ') v[16]', v, 64)
+    B2B_G(0, 8, 16, 24, SIGMA82[i * 16 + 0], SIGMA82[i * 16 + 1])
```
Comment (axic) [inline lib/evm/precompiles/09-blake2f.ts]: The round counter for `SIGMA` must be `mod 10`, `SIGMA82[(i % 10) * 16..]`. See the [RFC](https://tools.ietf.org/html/rfc7693#section-3.2).

Comment (s1na) [reply lib/evm/precompiles/09-blake2f.ts]: Wow, that was quick! Thanks, that fixed the failing test case :tada:

File: lib/evm/precompiles/09-blake2f.ts [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+export function F(h: Uint32Array, m: Uint32Array, t: Uint32Array, f: boolean, rounds: number) {
+  var i = 0
+
+  // init work variables
+  for (i = 0; i < 16; i++) {
+    v[i] = h[i]
+    v[i + 16] = BLAKE2B_IV32[i]
+  }
+
+  // low 64 bits of offset
```
Comment (axic) [inline lib/evm/precompiles/09-blake2f.ts]: Technically this should support all 128 bits of the offset. Unless [this change is accepted](https://github.com/ethereum/EIPs/issues/152#issuecomment-526645573) I'd suggest to keep an assertion here, that the upper 64-bits are 0.

File: lib/evm/precompiles/09-blake2f.ts [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+// Multiply them all by 2 to make them offsets into a uint32 buffer,
+// because this is Javascript and we don't have uint64s
+const SIGMA82 = new Uint8Array(
+  SIGMA8.map(function(x) {
+    return x * 2
+  }),
+)
+
+const v = new Uint32Array(32)
+var my = new Uint32Array(32)
```
Comment (axic) [inline lib/evm/precompiles/09-blake2f.ts]: Why are these outside?

**Comment by holgerd77** [comment]
That's great! 😄

@s1na Not sure, would likely make sense to release this as its own `blake2fjs` or so library so that others can benefit as well?

**Comment by holgerd77** [comment]
@s1na Ah, just did some reading on the EIP for the first time, so this (the F part) is just some extract from the `Blake 2b` hash function? First thought it would be an extra hash function on its own.

So eventually not useful to extract as an own lib? This part of the functionality is not directly callable through the API of the blakejs library?

File: lib/evm/precompiles/09-blake2f.ts (hunk 2 above)
Comment (s1na) [reply lib/evm/precompiles/09-blake2f.ts]: Sorry, that comment is misleading. The original code only supported the low 64 bits, but I modified it to take the whole 128 bits into account. Will fix the comment.

File: lib/evm/precompiles/09-blake2f.ts (hunk 3 above)
Comment (s1na) [reply lib/evm/precompiles/09-blake2f.ts]: Again, this is how blake2b.js did it. I tried not to change stuff much. I guess one reason is less memory allocations. The second var is also used as a global in `B2B_G`. I'll move them inside the function.

**Comment by s1na** [comment]
> Ah, just did some reading on the EIP for the first time, so this (the F part) is just some extract from the Blake 2b hash function?

Yeah exactly. It's kind of the core of blake2. You could e.g. build the blake2b hash function on top of this F function.

> So eventually not useful to extract as an own lib?

I'm not sure it'll be very useful. We could first release it here, and afterwards maybe move it to `ethereumjs-util` if we don't want to keep the code here.

> This part of the functionality is not directly callable through the API of the blakejs library?

Unfortunately not, I opened https://github.com/dcposch/blakejs/issues/12 a while back to ask whether that's an option, but haven't received a response. I'm not sure the repo is actively maintained.

File: lib/evm/precompiles/09-blake2f.ts [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+      exceptionError: new VmError(ERROR.OUT_OF_RANGE),
+    }
+  }
+
+  const rounds = data.slice(0, 4).readUInt32BE(0)
+  const hRaw = data.slice(4, 68)
+  const mRaw = data.slice(68, 196)
+  const tRaw = data.slice(196, 212)
+  // final
+  const f = data.slice(212, 213)[0] === 1
```
Comment (axic) [inline lib/evm/precompiles/09-blake2f.ts]: Could use `lastByte === 1` here.

File: lib/evm/precompiles/09-blake2f.ts [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+
+  const gasUsed = new BN(opts._common.param('gasPrices', 'blake2bRound'))
```
Comment (axic) [inline lib/evm/precompiles/09-blake2f.ts]: Nitpick, it should not be `blake2bRound` but `blake2Round`.

File: lib/evm/precompiles/09-blake2f.ts [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+  gasUsed.imul(new BN(rounds))
```
Comment (axic) [inline lib/evm/precompiles/09-blake2f.ts]: `imuln` given `rounds` in 32-bit?

File: lib/evm/precompiles/09-blake2f.ts (hunk 5 above)
Comment (s1na) [reply lib/evm/precompiles/09-blake2f.ts]: Created PR: https://github.com/ethereumjs/ethereumjs-common/pull/63

**Comment by holgerd77** [comment]
This is generally close-to-be-ready I would assume and we can wait to include this in the `v4.1` release?

**Comment by holgerd77** [comment]
@s1na Yes, likely makes sense to just leave the code here for now I would say.

**Comment by s1na** [comment]
> This is generally close-to-be-ready I would assume and we can wait to include this in the v4.1 release?

Yeah I don't have any more changes in mind, unless some more changes are requested during review.

Something to note is that Alex proposed some [changes](https://github.com/ethereum/EIPs/issues/152#issuecomment-526645573) to the EIP. But if they were merged in the EIP we could open a fresh PR later.

**Comment by axic** [comment]
@s1na this implementation seems to be very specific to blake2b and it may not work with blake2s (the G function needs to mod by the word with).

**Comment by holgerd77** [comment]
@axic Ah, the title is a bit misleading on the [EIP-152 PR page](https://github.com/ethereum/EIPs/pull/2129) - still stating `Blake2b`, also with similar references in the description - you should eventually update this (as it is in the final EIP text just stating `Blake2`), think this is a potential cause for a lot of confusion with first/superficial readers concluding this is an EIP for `Blake2b` support.

I would have a tendency to nevertheless merge here and release if the general review remarks are tackled - @s1na  or are these `Blake2s` related changes something which can be easily tested and modified? I think we will mark this `Istanbul` release anyhow as `beta` (so only the `Istanbul` support and not the VM release itself), seems there is still some discussion anyhow on some of the EIPs and this (all EIP implementations here) is also not tested yet against the official test suite. It's pretty likely that some things will further change along the way.

Does this make sense? The `Istanbul` release is pretty much awaited - together with some other fixes - so that people can start to test and integrate.

**Comment by axic** [comment]
It turns out the EIP is unclear and there is a debate going on. This PR matches the reference implementation done by the EIP authors and matches the tests given they were generated using that reference.

I hope there will be changes to the EIP, but that can happen in a subsequent PR here.

**Pull request merged**